texar/torch/data/tokenizers/sentencepiece_tokenizer.py

```diff
@@ -0,0 +1,256 @@
+    # pickled. We need to define __getstate__ here.
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        state["sp_model"] = None
+        state["vocab_file"] = None
+        return state, self.vocab_file
+
+    # spm.SentencePieceProcessor() is a SwigPyObject object which cannot be
+    # pickled. We need to define __setstate__ here.
+    def __setstate__(self, d):
+        self.__dict__, self.vocab_file = d
+        self.sp_model = spm.SentencePieceProcessor()
+        self.sp_model.Load(self.vocab_file)
+
+    def save_vocab(self, save_directory: str) -> Tuple[str]:
+        r"""Save the sentencepiece vocabulary (copy original file) to
+        a directory.
+        """
+        if not os.path.isdir(save_directory):
+            raise ValueError("Vocabulary path ({}) should be a "
+                             "directory".format(save_directory))
+        out_vocab_file = os.path.join(
+            save_directory, self._VOCAB_FILE_NAMES['vocab_file'])
+
+        if os.path.abspath(self.vocab_file) != os.path.abspath(out_vocab_file):
+            copyfile(self.vocab_file, out_vocab_file)
+
+        return (out_vocab_file,)
+
+    @property
+    def vocab_size(self) -> int:
+        return len(self.sp_model)
+
+    def _map_text_to_token(self, text: str) -> List[str]:  # type: ignore
+        return self.sp_model.EncodeAsPieces(text)
+
+    def _map_token_to_id(self, token: str) -> int:
+        return self.sp_model.PieceToId(token)
+
+    def _map_id_to_token(self, index: int) -> str:
+        token = self.sp_model.IdToPiece(index)
+        return token
+
+    def map_token_to_text(self, tokens: List[str]) -> str:
+        r"""Maps a sequence of tokens (string) in a single string."""
+        out_string = self.sp_model.DecodePieces(tokens)
+        return out_string
+
+    @staticmethod
+    def default_hparams() -> Dict[str, Any]:
+        r"""Returns a dictionary of hyperparameters with default values.
+
+        * The tokenizer is determined by `hparams['vocab_file']` if it's
+          specified. In this case, all other configurations in `hparams`
+          are ignored.
+        * Otherwise, the tokenizer is trained based on `hparams['text_file']`.
+        * `hparams['vocab_file']` and `hparams['text_file']` can not be None
+        at the same time.
+
+        .. code-block:: python
+
+            {
+                "vocab_file": None,
+                "text_file": None,
+                "vocab_size": None,
+                "model_type": "unigram",
+                "bos_token": "<s>",
+                "eos_token": "</s>",
+                "unk_token": "<unk>",
+                "pad_token": "<pad>",
+            }
+
+        Here:
+
+        `"vocab_file"`: str or None
+            The path to a sentencepiece vocabulary file.
+
+        `"text_file"`: str or None
+            Comma separated list of input sentences.
+
+        `"vocab_size"`: int or None
```

Comment:
What's the use of `vocab_size`? Will it have to be the same of the input vocab size? Or it's specified to truncate vocab? What's the effect of vocab_size == None? Pls make them clear in the docstring.

Reply:
`vocab_size` is used in the training procedure. You need to specify the `vocab_size` before you train it.

Reply:
pls add these details to docstring